vdaf/prio3/arith/templates/vector.go.tmpl

```diff
@@ -0,0 +1,233 @@
+// +build ignore
+// The previous line (and this one up to the warning below) is removed by the
+// template generator.
+
+// Code generated from ./templates/vector.go.tmpl. DO NOT EDIT.
+
+package {{.Name | ToLower}}
+
+import (
+	"io"
+	"encoding/binary"
+	"math/bits"
+
+	"github.com/cloudflare/circl/internal/conv"
+	"github.com/cloudflare/circl/internal/sha3"
+	"github.com/cloudflare/circl/math"
+	"golang.org/x/crypto/cryptobyte"
+)
+
+type Vec []Fp
+
+func (v Vec) Size() uint { return Size*uint(len(v)) }
+
+func (v Vec) AddAssign(x Vec) {
+	mustSameLen(v, x)
+	for i := range v {
+		v[i].AddAssign(&x[i])
+	}
+}
+
+func (v Vec) SubAssign(x Vec) {
+	mustSameLen(v, x)
+	for i := range v {
+		v[i].SubAssign(&x[i])
+	}
+}
+
+func (v Vec) ScalarMul(x *Fp) {
+	for i := range v {
+		v[i].MulAssign(x)
+	}
+}
+
+func (v Vec) DotProduct(x Vec) (out Fp) {
+	mustSameLen(v, x)
+	var t Fp
+	for i := range v {
+		t.Mul(&v[i], &x[i])
+		out.AddAssign(&t)
+	}
+	return
+}
+
+func bitRev(x uint, numBits uint) uint {
+	return bits.Reverse(x) >> (bits.UintSize - numBits)
+}
+
+func (v Vec) NTT(values Vec)    { v.doNTT(values, false) }
+func (v Vec) InvNTT(values Vec) { v.doNTT(values, true) }
+func (v Vec) doNTT(values Vec, isInvNTT bool) {
+	valuesLen := uint(len(values))
+	_, logN := math.NextPow2(valuesLen)
+
+	for i := range v {
+		j := bitRev(uint(i), logN)
+		if j < valuesLen {
+			v[i] = values[j]
+		}
+	}
+
+	var t, w, wn, iwn Fp
+	r := &wn
+	if isInvNTT {
+		r = &iwn
+		iwn.SetOne()
+	}
+
+	for l := uint(1); l <= logN; l++ {
+		y := uint(1) << (l - 1)
+		chunk := uint(1) << (logN - l)
+
+		for j := range chunk {
+			x := j << l
+			u := v[x]
+			v[x+0].Add(&u, &v[x+y])
+			v[x+y].Sub(&u, &v[x+y])
+		}
+
+		w.SetOne()
+		wn.SetRootOfUnityTwoN(l)
+		iwn.MulAssign(&wn)
+		for i := uint(1); i < y; i++ {
+			w.MulAssign(r)
+			for j := range chunk {
+				x := (j << l) + i
+				u := v[x]
+				t.Mul(&w, &v[x+y])
+				v[x+0].Add(&u, &t)
+				v[x+y].Sub(&u, &t)
+			}
+		}
+	}
+}
+
+func (v Vec) SplitBits(n uint64) error {
+	if bits.Len64(n) > len(v) {
+		return ErrNumberTooLarge
+	}
+
+	clear(v)
+	for i := range v {
+		if (n>>i)&0x1 == 1 {
+			v[i].SetOne()
+		}
+	}
+
+	return nil
+}
+
+func (v Vec) JoinBits() Fp {
+	var two Fp
+	_ = two.SetUint64(2)
+	return Poly(v).Evaluate(&two)
+}
+
+
+func (v Vec) Random(rnd io.Reader) error {
+	var b [Size]byte
+	var ok bool
+forVec:
+	for i := range v {
+		for range maxNumTries {
+			_, err := rnd.Read(b[:])
+			if err != nil {
+				return err
+			}
+
+			v[i], ok = isInRange(&b)
```

Comment:
I'm not sure you've implemented the masking properly: see [next_vec()](https://cfrg.github.io/draft-irtf-cfrg-vdaf/draft-irtf-cfrg-vdaf.html#section-6.2-6) in the draft. The idea is that if the modulus is 1 or more bits smaller than the `2 ** ENCODED_SIZE`, then we'll clear those higher bits before doing the range check.

This may not apply to either `Field64` or `Field128`. If not, we should at least document this in a way that makes it harder for us to make this mistake with different field parameters.

Reply:
correct, masking does not apply to these fields. 
I added a warning for future implementations.